src/main/java/teammates/common/util/RecaptchaVerifier.java

```diff
@@ -0,0 +1,86 @@
+package teammates.common.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URISyntaxException;
+import java.util.stream.Collectors;
+
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+/**
+ * Used to handle the verification of the user's reCAPTCHA response.
+ *
+ * @see <a href="https://developers.google.com/recaptcha/docs/verify">reCAPTCHA user response verification API</a>
+ * @see <a href="https://developers.google.com/recaptcha/docs/faq#id-like-to-run-automated-tests-with-recaptcha-what-should-i-do">Automated testing with reCAPTCHA v2</a>
+ */
+public final class RecaptchaVerifier {
+
+    /** The Google reCAPTCHA API URL to verify the response token. */
+    public static final String VERIFY_URL = "https://www.google.com/recaptcha/api/siteverify";
+
+    /** The shared secret key between the TEAMMATES site and reCAPTCHA. */
+    public static final String SECRET_KEY = Config.CAPTCHA_SECRET_KEY;
+
+    private static final Logger log = Logger.getLogger();
+
+    private RecaptchaVerifier() {
+        // utility class
+    }
+
+    /**
+     * Returns true if the {@code captchaResponse} token is verified successfully.
+     * @param captchaResponse The user's captcha response from the client side
+     * @return true if the response is verified successfully, and false if an exception occurs or if the request fails
+     */
+    public static boolean isVerificationSuccessful(String captchaResponse) {
+        if (captchaResponse == null || captchaResponse.isEmpty()) {
+            return false;
+        }
+
+        try {
+            URIBuilder urlb = new URIBuilder(VERIFY_URL);
+            urlb.setParameter("secret", SECRET_KEY);
+            urlb.setParameter("response", captchaResponse);
+
+            HttpPost post = new HttpPost(urlb.build());
+
+            //Execute and get the response.
+            try (CloseableHttpClient client = HttpClients.createDefault();
+                    CloseableHttpResponse resp = client.execute(post);
+                    BufferedReader br = new BufferedReader(new InputStreamReader(resp.getEntity().getContent()))) {
```

Comment:
Hmm I interpreted it a bit differently. I think it's suppressed only when both throw an exception.
> In contrast, in the example readFirstLineFromFile, if exceptions are thrown from **both** the try block and the try-with-resources statement, then the method readFirstLineFromFile throws the exception thrown from the try block; the exception thrown from the try-with-resources block is suppressed.
> An exception can be thrown from the block of code associated with the try-with-resources statement.

I'll try looking into it deeper later. But for now, I tried turning on airplane mode and ran the server in debug mode. It threw an `IOException` at line 58, which was caught at [line 77](https://github.com/TEAMMATES/teammates/pull/9447/files#diff-0249f1f8a6b5b317547bd6700e91200cR76). So, I think this will work. I'll just add a catch for `Exception` as well for safety.

On a side note, I used this implementation by looking at [`DatastoreBackupAction`](https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/ui/automated/DatastoreBackupAction.java#L64)

Reply:
> I think it's suppressed only when both throw an exception.

Yes this is what i meant too. 

As for u cutting off internet connection for testing, sometimes the fault happens on the third party server which you cannot control e.g. it returns you a null as the response or what so ever. So that's y i brought it up.

Reply:
Hmm yeah, so `resp.getEntity()` may throw `NullPointerException`.
From what I read ([here](https://javabeginnerstutorial.com/core-java-tutorial/exception-handling-try-resources/)), it will cause all the resources to close in reverse order. 

> Once the try-with-resources block completes (either successfully or by throwing an exception), the close methods of those resources will be invoked automatically in reverse order.

Also, I think the suppression ~~generally~~ happens only when there is an exception in the `close()` method of the resource. So we can ignore it in favour of the first exception from within the try **block**? 

In any case, `NullPointerException` isn't handled (and I don't think we can, due to a PMD rule).

So, I have 2 questions:
- Should I just put `Exception` in `catch` (to handle `null`)?
- Do we **need** to show the suppressed exception? If so, is it sufficient to use `e.getSuppressed()` or should I just scrap the `try-with-resources` statement entirely in favour of `try-catch-finally`?

Reply:
Yeah like what you brought up, anyways the null pointer exception is not handled. So what abt we handle it? as for the PMD error, you can take a look at AdminExceptionTestAction.java, theres a suppression msg that u can use. Yeah pls show the suppressed exception as well, not everyone knows there will be suppressed exceptions, so we can save some time on debugging. And as for the try-with-resources or try-catch-finally, you can choose the one with the cleaner implementation? just make sure the error is handled and ppl with no background knowledge on this can understand.